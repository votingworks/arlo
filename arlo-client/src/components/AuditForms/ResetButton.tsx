import React from 'react'
import ReactDOM from 'react-dom'
import { Button } from '@blueprintjs/core'
import { api } from '../utilities'

interface Props {
  updateAudit: () => void
  electionId: string
  disabled?: boolean
}

const ResetButton: React.FC<Props> = ({
  electionId,
  disabled,
  updateAudit,
}: Props) => {
  const resetButtonWrapper = document.getElementById('reset-button-wrapper')
  const reset = async () => {
    await api(`/audit/reset`, { electionId, method: 'POST' })

    updateAudit()
  }
  if (resetButtonWrapper) {
    return ReactDOM.createPortal(
      <Button onClick={reset} icon="refresh" disabled={disabled}>
        Clear &amp; Restart
      </Button>,
      resetButtonWrapper
    )
  }
  return null
}

export default ResetButton
