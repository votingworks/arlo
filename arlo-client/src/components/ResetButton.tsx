import React from 'react'
import ReactDOM from 'react-dom'
import { Button } from '@blueprintjs/core'
import { toast } from 'react-toastify'
import { api, checkAndToast } from './utilities'
import { IErrorResponse } from '../types'
import { useAuthDataContext } from './UserContext'

interface IProps {
  updateAudit: () => void
  electionId: string
  disabled?: boolean
}

const ResetButton: React.FC<IProps> = ({
  electionId,
  disabled,
  updateAudit,
}: IProps) => {
  const { isAuthenticated } = useAuthDataContext()
  const resetButtonWrapper = document.getElementById('reset-button-wrapper')
  const reset = async () => {
    try {
      const response: IErrorResponse = await api(
        `/election/${electionId}/audit/reset`,
        { method: 'POST' }
      )
      if (checkAndToast(response)) {
        return
      }
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    }
  }
  if (resetButtonWrapper && !isAuthenticated) {
    return ReactDOM.createPortal(
      <Button onClick={reset} icon="refresh" disabled={disabled}>
        Clear &amp; Restart
      </Button>,
      resetButtonWrapper
    )
  }
  return null
}

export default ResetButton
