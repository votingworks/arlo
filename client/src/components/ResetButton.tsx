import React from 'react'
import ReactDOM from 'react-dom'
import { Button } from '@blueprintjs/core'
import { api } from './utilities'
import { IErrorResponse } from '../types'
import { useAuthDataContext } from './UserContext'

interface IProps {
  updateAudit: () => void
  electionId: string
  disabled?: boolean
}

const ResetButton: React.FC<IProps> = ({
  electionId,
  disabled,
  updateAudit,
}: IProps) => {
  const { isAuthenticated } = useAuthDataContext()
  const resetButtonWrapper = document.getElementById('reset-button-wrapper')
  const reset = async () => {
    const response = await api<IErrorResponse>(
      `/election/${electionId}/audit/reset`,
      { method: 'POST' }
    )
    if (!response) {
      return
    }
    updateAudit()
  }
  if (resetButtonWrapper && !isAuthenticated) {
    return ReactDOM.createPortal(
      <Button onClick={reset} icon="refresh" disabled={disabled}>
        Clear &amp; Restart
      </Button>,
      resetButtonWrapper
    )
  }
  return null
}

export default ResetButton
