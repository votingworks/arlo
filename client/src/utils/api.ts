import { AxiosRequestConfig } from 'axios'

const parseCookies = () =>
  Object.fromEntries(
    document.cookie.split(';').map(pair => pair.trim().split('='))
  )

export const addCSRFToken = (options?: RequestInit | AxiosRequestConfig) => {
  const token = parseCookies()._csrf_token
  if (
    token &&
    options &&
    ['POST', 'PUT', 'PATCH', 'DELETE'].includes(options.method!)
  )
    return {
      ...options,
      headers: { ...options.headers, 'X-CSRFToken': token },
    }
  return options
}

export const tryJson = (responseText: string) => {
  try {
    return JSON.parse(responseText)
  } catch (err) {
    return {}
  }
}
export class ApiError extends Error {
  public statusCode: number

  public constructor(message: string, statusCode: number) {
    super(message)
    this.statusCode = statusCode
  }
}

export const fetchApi = async (url: string, options?: RequestInit) => {
  const response = await fetch(url, addCSRFToken(options))
  if (response.ok) return response.json()
  const text = await response.text()
  const { errors } = tryJson(text)
  const error = errors && errors.length && errors[0].message
  throw new ApiError(error || text, response.status)
}
