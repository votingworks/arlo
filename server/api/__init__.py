from flask import Blueprint

api = Blueprint("api", __name__)

# pylint: disable=wrong-import-position,cyclic-import

# Single-jurisdiction flow routes
from . import routes

# Multi-jurisdiction flow routes
from . import election_settings
from . import contests
from . import jurisdictions
from . import standardized_contests
from . import rounds
from . import sample_sizes
from . import ballot_manifest
from . import batch_tallies
from . import cvrs
from . import audit_boards
from . import ballots
from . import batches
from . import offline_results
from . import reports
