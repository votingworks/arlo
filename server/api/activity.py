from flask import session
from flask.json import jsonify
from werkzeug.exceptions import Forbidden

from ..models import *  # pylint: disable=wildcard-import
from ..auth.lib import UserType, get_loggedin_user
from . import api
from ..util.isoformat import isoformat


@api.route("/organizations/<organization_id>/activities", methods=["GET"])
def list_activities(organization_id: str):
    user_type, user_key = get_loggedin_user(session)
    user = User.query.filter_by(email=user_key).one_or_none()
    if (
        not user
        or user_type != UserType.AUDIT_ADMIN
        or not any(org.id == organization_id for org in user.organizations)
    ):
        return Forbidden()

    activities = (
        ActivityLogRecord.query.filter_by(organization_id=organization_id)
        .order_by(ActivityLogRecord.timestamp.desc())
        .all()
    )
    return jsonify(
        [
            dict(
                id=activity.id,
                activityName=activity.activity_name,
                timestamp=isoformat(activity.timestamp),
                user=(
                    activity.info["base"].get("user_key")
                    and dict(
                        key=activity.info["base"]["user_key"],
                        type=activity.info["base"]["user_type"],
                        supportUser=(activity.info["base"].get("support_user_email")),
                    )
                ),
                election=(
                    activity.info["base"].get("election_id")
                    and dict(
                        id=activity.info["base"]["election_id"],
                        auditName=activity.info["base"]["audit_name"],
                        auditType=activity.info["base"]["audit_type"],
                    )
                ),
                info={key: val for key, val in activity.info.items() if key != "base"},
            )
            for activity in activities
        ]
    )
