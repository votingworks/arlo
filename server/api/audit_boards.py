import uuid
from datetime import datetime
from typing import List, Dict
from flask import jsonify, request
from xkcdpass import xkcd_password as xp
from werkzeug.exceptions import Conflict, BadRequest
from sqlalchemy import func, and_
from sqlalchemy.orm import contains_eager

from ..app import app, db
from ..models import *  # pylint: disable=wildcard-import
from ..auth import with_jurisdiction_access, with_audit_board_access
from .rounds import get_current_round
from .sample_sizes import cumulative_contest_results
from ..util.jsonschema import validate, JSONDict
from ..util.binpacking import BalancedBucketList, Bucket
from ..util.group_by import group_by
from ..util.isoformat import isoformat
from ..audit_math import bravo, sampler_contest

WORDS = xp.generate_wordlist(wordfile=xp.locate_wordfile())

CREATE_AUDIT_BOARD_REQUEST_SCHEMA = {
    "type": "object",
    "properties": {"name": {"type": "string"},},
    "additionalProperties": False,
    "required": ["name"],
}

# Raises if invalid
def validate_audit_boards(
    audit_boards: List[JSONDict],
    election: Election,
    jurisdiction: Jurisdiction,
    round: Round,
):
    current_round = get_current_round(election)
    if not current_round or round.id != current_round.id:
        raise Conflict(f"Round {round.round_num} is not the current round")

    if any(ab for ab in jurisdiction.audit_boards if ab.round_id == round.id):
        raise Conflict(f"Audit boards already created for round {round.round_num}")

    validate(
        audit_boards, {"type": "array", "items": CREATE_AUDIT_BOARD_REQUEST_SCHEMA}
    )

    if len(set(ab["name"] for ab in audit_boards)) != len(audit_boards):
        raise BadRequest("Audit board names must be unique")


def assign_sampled_ballots(
    jurisdiction: Jurisdiction, round: Round, audit_boards: List[AuditBoard],
):
    # Collect the physical ballots for each batch that were sampled for this
    # jurisdiction for this round
    sampled_ballots = (
        SampledBallot.query.join(Batch)
        .filter_by(jurisdiction_id=jurisdiction.id)
        .join(SampledBallot.draws)
        .filter_by(round_id=round.id)
        .order_by(Batch.name)  # group_by prefers a sorted list
        .options(contains_eager(SampledBallot.batch))
        .all()
    )
    ballots_by_batch = group_by(sampled_ballots, key=lambda sb: sb.batch.name)

    # Divvy up batches of ballots between the audit boards.
    # Note: BalancedBucketList doesn't care which buckets have which batches to
    # start, so we add all the batches to the first bucket before balancing.
    buckets = [Bucket(audit_board.id) for audit_board in audit_boards]
    for batch_name, sampled_ballots in ballots_by_batch.items():
        buckets[0].add_batch(batch_name, len(sampled_ballots))
    balanced_buckets = BalancedBucketList(buckets)

    for bucket in balanced_buckets.buckets:
        ballots_in_bucket = [
            ballot
            for batch_name in bucket.batches
            for ballot in ballots_by_batch[batch_name]
        ]
        for ballot in ballots_in_bucket:
            ballot.audit_board_id = bucket.name
            db.session.add(ballot)


@app.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/audit-board",
    methods=["POST"],
)
@with_jurisdiction_access
def create_audit_boards(election: Election, jurisdiction: Jurisdiction, round_id: str):
    json_audit_boards = request.get_json()
    round = Round.query.get_or_404(round_id)
    validate_audit_boards(json_audit_boards, election, jurisdiction, round)

    audit_boards = [
        AuditBoard(
            id=str(uuid.uuid4()),
            name=json_audit_board["name"],
            jurisdiction_id=jurisdiction.id,
            round_id=round.id,
            passphrase=xp.generate_xkcdpassword(WORDS, numwords=4, delimiter="-"),
        )
        for json_audit_board in json_audit_boards
    ]
    db.session.add_all(audit_boards)

    assign_sampled_ballots(jurisdiction, round, audit_boards)

    db.session.commit()

    return jsonify(status="ok")


def round_status_by_audit_board(
    jurisdiction_id: str, round_id: str
) -> Dict[str, JSONDict]:
    audit_boards = AuditBoard.query.filter_by(
        jurisdiction_id=jurisdiction_id, round_id=round_id
    ).all()
    sampled_ballots_by_audit_board = dict(
        SampledBallot.query.join(AuditBoard)
        .filter_by(jurisdiction_id=jurisdiction_id)
        .group_by(AuditBoard.id)
        .values(AuditBoard.id, func.count())
    )
    audited_ballots_by_audit_board = dict(
        SampledBallot.query.join(AuditBoard)
        .filter_by(jurisdiction_id=jurisdiction_id)
        .filter(SampledBallot.status != BallotStatus.NOT_AUDITED)
        .group_by(AuditBoard.id)
        .values(AuditBoard.id, func.count())
    )

    return {
        ab.id: {
            "numSampledBallots": sampled_ballots_by_audit_board.get(ab.id, 0),
            "numAuditedBallots": audited_ballots_by_audit_board.get(ab.id, 0),
        }
        for ab in audit_boards
    }


def serialize_audit_board(audit_board: AuditBoard, round_status: JSONDict) -> JSONDict:
    return {
        "id": audit_board.id,
        "name": audit_board.name,
        "passphrase": audit_board.passphrase,
        "signedOffAt": isoformat(audit_board.signed_off_at),
        "currentRoundStatus": round_status,
    }


def serialize_members(audit_board):
    members = []

    for i in range(0, 2):
        name = getattr(audit_board, f"member_{i + 1}")
        affiliation = getattr(audit_board, f"member_{i + 1}_affiliation")

        if not name:
            break

        members.append({"name": name, "affiliation": affiliation})

    return members


@app.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/audit-board",
    methods=["GET"],
)
@with_jurisdiction_access
def list_audit_boards(
    election: Election,  # pylint: disable=unused-argument
    jurisdiction: Jurisdiction,
    round_id: str,
):
    Round.query.get_or_404(round_id)
    audit_boards = (
        AuditBoard.query.filter_by(jurisdiction_id=jurisdiction.id, round_id=round_id)
        .order_by(AuditBoard.name)
        .all()
    )
    round_status = round_status_by_audit_board(jurisdiction.id, round_id)
    json_audit_boards = [
        serialize_audit_board(ab, round_status[ab.id]) for ab in audit_boards
    ]
    return jsonify({"auditBoards": json_audit_boards})


MEMBER_SCHEMA = {
    "type": "object",
    "properties": {
        "name": {"type": "string"},
        "affiliation": {
            "anyOf": [
                {
                    "type": "string",
                    "enum": [affiliation.value for affiliation in Affiliation],
                },
                {"type": "null"},
            ]
        },
    },
    "additionalProperties": False,
    "required": ["name", "affiliation"],
}

SET_MEMBERS_SCHEMA = {
    "type": "array",
    "items": MEMBER_SCHEMA,
}


def validate_members(members: List[JSONDict]):
    # You can do all of these checks using JSON schema, but the resulting error
    # messages aren't very good.
    if len(members) == 0:
        raise BadRequest("Must have at least one member.")
    if len(members) > 2:
        raise BadRequest("Cannot have more than two members.")

    validate(members, SET_MEMBERS_SCHEMA)

    for member in members:
        if member["name"] == "":
            raise BadRequest("'name' must not be empty.")


@app.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/audit-board/<audit_board_id>/members",
    methods=["PUT"],
)
@with_audit_board_access
def set_audit_board_members(
    election: Election,  # pylint: disable=unused-argument
    jurisdiction: Jurisdiction,  # pylint: disable=unused-argument
    round: Round,  # pylint: disable=unused-argument
    audit_board: AuditBoard,
):
    members = request.get_json()
    validate_members(members)

    audit_board.member_1 = members[0]["name"]
    audit_board.member_1_affiliation = members[0]["affiliation"]
    if len(members) > 1:
        audit_board.member_2 = members[1]["name"]
        audit_board.member_2_affiliation = members[1]["affiliation"]

    db.session.commit()

    return jsonify(status="ok")


def calculate_risk_measurements(election: Election, round: Round):
    for contest in election.contests:
        risk, is_complete = bravo.compute_risk(
            election.risk_limit / 100,
            sampler_contest.from_db_contest(contest),
            cumulative_contest_results(contest),
        )

        round_contest = next(
            rc for rc in round.round_contests if rc.contest_id == contest.id
        )
        round_contest.end_p_value = max(risk.values())
        round_contest.is_complete = is_complete


def count_audited_votes(election: Election, round: Round):
    vote_counts = dict(
        BallotInterpretation.query.filter_by(is_overvote=False)
        .join(SampledBallot)
        .join(SampledBallotDraw)
        .filter_by(round_id=round.id)
        .join(BallotInterpretation.selected_choices)
        .group_by(ContestChoice.id)
        .values(ContestChoice.id, func.count())
    )

    for contest in election.contests:
        for contest_choice in contest.choices:
            result = RoundContestResult(
                round_id=round.id,
                contest_id=contest.id,
                contest_choice_id=contest_choice.id,
                result=vote_counts.get(contest_choice.id, 0),
            )
            db.session.add(result)


def end_round(election: Election, round: Round):
    count_audited_votes(election, round)
    calculate_risk_measurements(election, round)
    round.ended_at = datetime.utcnow()


def is_round_complete(election: Election, round: Round) -> bool:
    num_jurisdictions_without_audit_boards_set_up: int = (
        # For each jurisdiction...
        Jurisdiction.query.filter_by(election_id=election.id)
        # Where there are ballots that haven't been audited...
        .join(Jurisdiction.batches)
        .join(Batch.ballots)
        .filter(SampledBallot.status == BallotStatus.NOT_AUDITED)
        # And those ballots got sampled this round...
        .join(SampledBallot.draws)
        .filter_by(round_id=round.id)
        # Count the number of audit boards set up.
        .outerjoin(
            AuditBoard,
            and_(
                AuditBoard.jurisdiction_id == Jurisdiction.id,
                AuditBoard.round_id == round.id,
            ),
        )
        .group_by(Jurisdiction.id)
        # Finally, count how many jurisdictions have no audit boards set up.
        .having(func.count(AuditBoard.id) == 0)
        .count()
    )
    num_audit_boards_with_ballots_not_signed_off: int = (
        AuditBoard.query.filter_by(round_id=round.id, signed_off_at=None)
        .join(SampledBallot)
        .group_by(AuditBoard.id)
        .count()
    )
    return (
        num_jurisdictions_without_audit_boards_set_up == 0
        and num_audit_boards_with_ballots_not_signed_off == 0
    )


SIGN_OFF_AUDIT_BOARD_REQUEST_SCHEMA = {
    "type": "object",
    "properties": {
        "memberName1": {"type": "string"},
        "memberName2": {"type": "string"},
    },
    "additionalProperties": False,
    "required": ["memberName1", "memberName2"],
}

# Raises if invalid
def validate_sign_off(sign_off_request: JSONDict, audit_board: AuditBoard):
    validate(sign_off_request, SIGN_OFF_AUDIT_BOARD_REQUEST_SCHEMA)

    if sign_off_request["memberName1"] != audit_board.member_1:
        raise BadRequest(
            f"Audit board member name did not match: {sign_off_request['memberName1']}"
        )

    if audit_board.member_2 and sign_off_request["memberName2"] != audit_board.member_2:
        raise BadRequest(
            f"Audit board member name did not match: {sign_off_request['memberName2']}"
        )

    if any(b.status == BallotStatus.NOT_AUDITED for b in audit_board.sampled_ballots):
        raise Conflict("Audit board is not finished auditing all assigned ballots")


@app.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/audit-board/<audit_board_id>/sign-off",
    methods=["POST"],
)
@with_audit_board_access
def sign_off_audit_board(
    election: Election,
    jurisdiction: Jurisdiction,  # pylint: disable=unused-argument
    round: Round,
    audit_board: AuditBoard,
):
    validate_sign_off(request.get_json(), audit_board)

    audit_board.signed_off_at = datetime.utcnow()

    if is_round_complete(election, round):
        end_round(election, round)

    db.session.commit()

    return jsonify(status="ok")
