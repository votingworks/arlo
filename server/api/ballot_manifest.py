import locale, uuid
from datetime import datetime
from sqlalchemy.orm.session import Session
from flask import request, jsonify, Request
from werkzeug.exceptions import BadRequest, NotFound

from . import api
from ..models import *  # pylint: disable=wildcard-import
from ..auth import with_jurisdiction_access, with_election_access
from ..util.process_file import (
    process_file,
    serialize_file,
    serialize_file_processing,
)
from ..util.csv_download import csv_response
from ..util.csv_parse import decode_csv_file, parse_csv, CSVValueType, CSVColumnType

BATCH_NAME = "Batch Name"
NUMBER_OF_BALLOTS = "Number of Ballots"
STORAGE_LOCATION = "Storage Location"
TABULATOR = "Tabulator"

BALLOT_MANIFEST_COLUMNS = [
    CSVColumnType(BATCH_NAME, CSVValueType.TEXT, unique=True),
    CSVColumnType(NUMBER_OF_BALLOTS, CSVValueType.NUMBER),
    CSVColumnType(STORAGE_LOCATION, CSVValueType.TEXT, required=False),
    CSVColumnType(TABULATOR, CSVValueType.TEXT, required=False),
]


def process_ballot_manifest_file(
    session: Session, jurisdiction: Jurisdiction, file: File
):
    assert jurisdiction.manifest_file_id == file.id

    def process():
        manifest_csv = parse_csv(
            jurisdiction.manifest_file.contents, BALLOT_MANIFEST_COLUMNS
        )

        num_batches = 0
        num_ballots = 0
        for row in manifest_csv:
            num_ballots_in_batch_csv = row[NUMBER_OF_BALLOTS]
            num_ballots_in_batch = locale.atoi(num_ballots_in_batch_csv)

            batch = Batch(
                id=str(uuid.uuid4()),
                name=row[BATCH_NAME],
                jurisdiction_id=jurisdiction.id,
                num_ballots=num_ballots_in_batch,
                storage_location=row.get(STORAGE_LOCATION) or None,
                tabulator=row.get(TABULATOR) or None,
            )
            session.add(batch)
            num_batches += 1
            num_ballots += batch.num_ballots

        jurisdiction.manifest_num_ballots = num_ballots
        jurisdiction.manifest_num_batches = num_batches

    process_file(session, file, process)

    election = jurisdiction.election

    # If we're in the single-jurisdiction flow, posting the ballot manifest
    # starts the first round, so we need to sample the ballots.
    # In the multi-jurisdiction flow, this happens after all jurisdictions
    # upload manifests, and is triggered by a different endpoint.
    if not election.is_multi_jurisdiction:
        # Import this here to avoid circular dependencies
        # pylint: disable=import-outside-toplevel,cyclic-import
        from .routes import sample_ballots

        sample_ballots(session, election, election.rounds[0])


# Raises if invalid
def validate_ballot_manifest_upload(request: Request):
    if "manifest" not in request.files:
        raise BadRequest("Missing required file parameter 'manifest'")


# We save the ballot manifest file, and bgcompute finds it and processes it in
# the background.
def save_ballot_manifest_file(manifest, jurisdiction: Jurisdiction):
    manifest_string = decode_csv_file(manifest.read())
    jurisdiction.manifest_file = File(
        id=str(uuid.uuid4()),
        name=manifest.filename,
        contents=manifest_string,
        uploaded_at=datetime.utcnow(),
    )


def clear_ballot_manifest_file(jurisdiction: Jurisdiction):
    jurisdiction.manifest_num_ballots = None
    jurisdiction.manifest_num_batches = None

    if jurisdiction.manifest_file_id:
        File.query.filter_by(id=jurisdiction.manifest_file_id).delete()
    Batch.query.filter_by(jurisdiction=jurisdiction).delete()


@api.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/ballot-manifest",
    methods=["PUT"],
)
@with_jurisdiction_access
def upload_ballot_manifest(
    election: Election, jurisdiction: Jurisdiction,  # pylint: disable=unused-argument
):
    validate_ballot_manifest_upload(request)
    clear_ballot_manifest_file(jurisdiction)
    save_ballot_manifest_file(request.files["manifest"], jurisdiction)
    db.session.commit()
    return jsonify(status="ok")


@api.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/ballot-manifest",
    methods=["GET"],
)
@with_jurisdiction_access
def get_ballot_manifest(
    election: Election, jurisdiction: Jurisdiction  # pylint: disable=unused-argument
):
    if jurisdiction.manifest_file:
        return jsonify(
            file=serialize_file(jurisdiction.manifest_file),
            processing=serialize_file_processing(jurisdiction.manifest_file),
        )
    else:
        return jsonify(file=None, processing=None)


@api.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/ballot-manifest/csv",
    methods=["GET"],
)
@with_election_access
def download_ballot_manifest_file(
    election: Election, jurisdiction_id: str,  # pylint: disable=unused-argument
):
    jurisdiction = Jurisdiction.query.filter_by(
        election_id=election.id, id=jurisdiction_id
    ).first()
    if not jurisdiction or not jurisdiction.manifest_file:
        return NotFound()

    return csv_response(
        jurisdiction.manifest_file.contents, jurisdiction.manifest_file.name
    )


@api.route(
    "/election/<election_id>/jurisdiction/<jurisdiction_id>/ballot-manifest",
    methods=["DELETE"],
)
@with_jurisdiction_access
def clear_ballot_manifest(
    election: Election, jurisdiction: Jurisdiction,  # pylint: disable=unused-argument
):
    clear_ballot_manifest_file(jurisdiction)
    db.session.commit()
    return jsonify(status="ok")
