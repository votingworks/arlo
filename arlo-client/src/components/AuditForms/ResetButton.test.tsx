import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import ResetButton from './ResetButton'
import api from '../utilities'

const apiMock = api as jest.Mock<ReturnType<typeof api>, Parameters<typeof api>>

jest.mock('../utilities')
apiMock.mockImplementationOnce(async () => ({}))

describe('ResetButton', () => {
  it('renders', () => {
    const updateAuditMock = jest.fn()
    const { container } = render(
      <ResetButton electionId="1" updateAudit={updateAuditMock} />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders disabled', () => {
    const updateAuditMock = jest.fn()
    const { container } = render(
      <ResetButton electionId="1" updateAudit={updateAuditMock} />
    )
    expect(container).toMatchSnapshot()
  })

  it('posts to /audit/reset and calls updateAudit', async () => {
    const updateAuditMock = jest.fn()
    const wrapper = document.createElement('div')
    wrapper.setAttribute('id', 'reset-button-wrapper')

    const { getByText } = render(
      <ResetButton electionId="1" updateAudit={updateAuditMock} />,
      { container: document.body.appendChild(wrapper) }
    )

    fireEvent.click(getByText('Clear & Restart'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toHaveBeenCalledTimes(1)
      expect(updateAuditMock).toHaveBeenCalledTimes(1)
    })
  })
})
