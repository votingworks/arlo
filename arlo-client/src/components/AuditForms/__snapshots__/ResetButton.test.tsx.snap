// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetButton renders 1`] = `<div />`;

exports[`ResetButton renders disabled 1`] = `<div />`;
